### PR TITLE
TIP-694: rework the completeness checkers

### PR DESCRIPTION
**Pull request by jjanvier**
## This PR about the completeness checkers

For our rework of the completeness, we'll try to use the completeness checkers to determine if a value is complete or not.

This PR cleans the checkers, handles cases that were not covered and adds specs. It also changes the API to make it explicit. Now, the given channel and locale are taken into account to know if the checkers supports the value or not. Before this PR, the checkers didn't check if a value was complete. Actually they checked if the data of the value was filled in or not.

## Global completeness rework plan

We still don't know if this approach (calculation with POO only) is performant enough, but for sure it allows us to write a quick and simple version of the completeness calculation. Which means, in the coming PRs, we will:

- add some tests on the completeness (see https://github.com/akeneo/pim-community-dev/pull/5775)
- fix/enhance the completeness checkers (this PR)
- add other tests that we couldn't add on master
- write a first naive and simple version of the completeness calculation (with the tests green)
- benchmark
- [depending of the results of the benchmark, maybe trash the simple code to test another path]

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Y
| Added Behats                      | -
| Added integration tests           | -
| Changelog updated                 | Y
| Review and 2 GTM                  | 
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

### CODE REVIEW

File: src/Pim/Component/Catalog/Completeness/Checker/ChainedProductValueCompleteChecker.php [hunk 1]
```diff
@@ -7,9 +7,14 @@
...
 
 /**
+ * Chained checker that contains all the product value completeness checkers.
+ *
  * @author    JM Leroux <jean-marie.leroux@akeneo.com>
  * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
  * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ *
+ * @internal for internal use only, please use the \Pim\Component\Catalog\Completeness\CompletenessGeneratorInterface
+ *           to calculate the completeness on a product
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Completeness/Checker/ChainedProductValueCompleteChecker.php]: :+1:

File: src/Pim/Component/Catalog/Completeness/Checker/PriceCompleteChecker.php [hunk 2]
```diff
@@ -43,6 +58,26 @@ function ($currency) {
...
+
+    /**
+     * Get the list of currency codes to check in the product value.
+     *
+     * @param AttributeInterface    $attribute
+     * @param ChannelInterface|null $channel
+     *
+     * @return array
+     */
+    protected function currencyCodesToCheck(AttributeInterface $attribute, ChannelInterface $channel = null)
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Completeness/Checker/PriceCompleteChecker.php]: `getCurrencyCodesToCheck`

**Review by damien-carcel — APPROVED** [review]

File: src/Pim/Component/Catalog/Completeness/Checker/MetricCompleteChecker.php [hunk 3]
```diff
@@ -35,6 +53,6 @@ public function isComplete(
      */
     public function supportsValue(ProductValueInterface $productValue)
     {
-        return 'pim_catalog_metric' === $productValue->getAttribute()->getAttributeType();
+        return AttributeTypes::METRIC === $productValue->getAttribute()->getAttributeType();
```
Comment (SamirBoulil) [inline src/Pim/Component/Catalog/Completeness/Checker/MetricCompleteChecker.php]: 👍

File: src/Pim/Component/Catalog/Completeness/Checker/PriceCompleteChecker.php [hunk 4]
```diff
@@ -43,6 +58,26 @@ function ($currency) {
      */
     public function supportsValue(ProductValueInterface $productValue)
```
Comment (SamirBoulil) [inline src/Pim/Component/Catalog/Completeness/Checker/PriceCompleteChecker.php]: I'd rename this for `supportsProductValue`, but it's ok

**Review by SamirBoulil — APPROVED** [review]
Seems cool, and will be helpful for the next steps !

What do you think of having those implementations in the ProductValue concrete classes or even being decorated by a class which does this. 

Could be cool to have a ProductValue capable of telling if it is complete (that way having a munch capable classes).

**Comment by jjanvier** [comment]
seen IRL with @SamirBoulil why it's not yet possible for a product value to determine itself if it's complete or not. Basically, it's currently impossible because of the prices :(  

Could be possible once IM-758 fixed.

**Review by SamirBoulil — APPROVED** [review]

File: src/Pim/Bundle/CatalogBundle/Resources/config/completeness_checkers.yml [hunk 5]
```diff
@@ -3,28 +3,33 @@ parameters:
     pim_catalog.completeness.checker.media.class:   Pim\Component\Catalog\Completeness\Checker\MediaCompleteChecker
     pim_catalog.completeness.checker.metric.class:  Pim\Component\Catalog\Completeness\Checker\MetricCompleteChecker
     pim_catalog.completeness.checker.price.class:   Pim\Component\Catalog\Completeness\Checker\PriceCompleteChecker
-    pim_catalog.completeness.checker.empty.class:   Pim\Component\Catalog\Completeness\Checker\EmptyChecker
+    pim_catalog.completeness.checker.simple.class:  Pim\Component\Catalog\Completeness\Checker\SimpleCompleteChecker
 
 services:
     pim_catalog.completeness.checker.chained:
```
Comment (arnolanglade) [inline src/Pim/Bundle/CatalogBundle/Resources/config/completeness_checkers.yml]: We should call it `Pim\Component\Catalog\Completeness\Checker\ProductValueCompleteChecker` and its key should be `pim_catalog.completeness.checker`

**Review by damien-carcel — APPROVED** [review]

File: src/Pim/Bundle/CatalogBundle/Resources/config/completeness_checkers.yml [hunk 6]
```diff
@@ -3,28 +3,33 @@ parameters:
...
             - { name: pim_catalog.completeness.checker.product_value, priority: 100 }
 
     pim_catalog.completeness.checker.price:
         class: '%pim_catalog.completeness.checker.price.class%'
+        public: false
         tags:
             - { name: pim_catalog.completeness.checker.product_value, priority: 100 }
+
+    pim_catalog.completeness.checker.simple:
+        class: '%pim_catalog.completeness.checker.simple.class%'
```
Comment (arnolanglade) [inline src/Pim/Bundle/CatalogBundle/Resources/config/completeness_checkers.yml:31]: I don't like "simple", is it a PIM convention naming? does it only check completeness for scalar value?

File: src/Pim/Component/Catalog/Completeness/Checker/ChainedProductValueCompleteChecker.php (hunk 1 above)
Comment (arnolanglade) [reply src/Pim/Component/Catalog/Completeness/Checker/ChainedProductValueCompleteChecker.php]: should this class be final? :trollface:

File: src/Pim/Component/Catalog/Completeness/Checker/ChainedProductValueCompleteChecker.php [hunk 7]
```diff
@@ -21,32 +31,43 @@ class ChainedProductValueCompleteChecker implements ProductValueCompleteCheckerI
      */
     public function isComplete(
         ProductValueInterface $productValue,
-        ChannelInterface $channel = null,
-        LocaleInterface $locale = null
+        ChannelInterface $channel,
+        LocaleInterface $locale
     ) {
         foreach ($this->productValueCheckers as $productValueChecker) {
```
Comment (arnolanglade) [inline src/Pim/Component/Catalog/Completeness/Checker/ChainedProductValueCompleteChecker.php]: Should be :

```php
foreach ($this->productValueCheckers as $productValueChecker) {
    if ($productValueChecker->supportsValue($productValue, $channel, $locale)) {
         return $productValueChecker->isComplete($productValue, $channel, $locale); // it breack the loop if the check is supported.
    }
}
return false;
```

File: src/Pim/Component/Catalog/Completeness/Checker/ChainedProductValueCompleteChecker.php [hunk 8]
```diff
@@ -21,32 +31,43 @@ class ChainedProductValueCompleteChecker implements ProductValueCompleteCheckerI
...
+            return false;
+        }
+
+        if ($productValue->getAttribute()->isLocaleSpecific() &&
+            !$productValue->getAttribute()->hasLocaleSpecific($locale)
+        ) {
+            return false;
+        }
+
+        return true;
```
Comment (arnolanglade) [inline src/Pim/Component/Catalog/Completeness/Checker/ChainedProductValueCompleteChecker.php]: Extra question: should we need `addProductValueChecker`? Should we inject them in the constructor?

File: src/Pim/Component/Catalog/Completeness/Checker/PriceCompleteChecker.php [hunk 9]
```diff
@@ -41,8 +43,24 @@ function ($currency) {
...
+    ) {
+        return AttributeTypes::PRICE_COLLECTION === $productValue->getAttribute()->getAttributeType();
+    }
+
+    /**
+     * @param ChannelInterface|null $channel
+     *
+     * @return array
+     */
+    private function getCurrencyCodesToCheck(ChannelInterface $channel)
```
Comment (arnolanglade) [inline src/Pim/Component/Catalog/Completeness/Checker/PriceCompleteChecker.php]: :heart:

File: src/Pim/Component/Catalog/Completeness/Checker/PriceCompleteChecker.php [hunk 10]
```diff
@@ -41,8 +43,24 @@ function ($currency) {
...
     {
-        return 'pim_catalog_price_collection' === $productValue->getAttribute()->getAttributeType();
+        return array_map(
+            function ($currency) { return $currency->getCode(); },
+            $channel->getCurrencies()->toArray()
```
Comment (arnolanglade) [inline src/Pim/Component/Catalog/Completeness/Checker/PriceCompleteChecker.php]: Should be `$channel->getCurrencies()->map(function() {})`

File: src/Pim/Component/Catalog/Completeness/Checker/SimpleCompleteChecker.php [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    public function supportsValue(
+        ProductValueInterface $productValue,
+        ChannelInterface $channel,
+        LocaleInterface $locale
+    ) {
+        return true;
+    }
+
+    /**
+     * @param $data
```
Comment (arnolanglade) [inline src/Pim/Component/Catalog/Completeness/Checker/SimpleCompleteChecker.php]: mixed $data

File: src/Pim/Component/Catalog/Completeness/Checker/ChainedProductValueCompleteChecker.php [hunk 12]
```diff
@@ -21,32 +31,43 @@ class ChainedProductValueCompleteChecker implements ProductValueCompleteCheckerI
...
     /**
      * {@inheritdoc}
      */
-    public function supportsValue(ProductValueInterface $productValue)
-    {
-        foreach ($this->productValueCheckers as $productValueChecker) {
-            if ($productValueChecker->supportsValue($productValue)) {
-                return true;
-            }
+    public function supportsValue(
```
Comment (arnolanglade) [inline src/Pim/Component/Catalog/Completeness/Checker/ChainedProductValueCompleteChecker.php]: This method should be empty and return true. Those checks should be in the `isComplete` method.

File: src/Pim/Bundle/CatalogBundle/Resources/config/completeness_checkers.yml (hunk 5 above)
Comment (jjanvier) [reply src/Pim/Bundle/CatalogBundle/Resources/config/completeness_checkers.yml]: as seen IRL, `Pim\Component\Catalog\Completeness\Checker\ProductValueCompleteChecker` will be the used for the chained checker, as it's the main entry point

File: src/Pim/Bundle/CatalogBundle/Resources/config/completeness_checkers.yml (hunk 6 above)
Comment (jjanvier) [reply src/Pim/Bundle/CatalogBundle/Resources/config/completeness_checkers.yml:31]: it's checks all product values except metrics, prices and medias

yes, it's mainly a convention
"default" would be better imo

Comment (jjanvier) [reply src/Pim/Bundle/CatalogBundle/Resources/config/completeness_checkers.yml:31]: (seen IRL, we'll keep simple)

File: src/Pim/Component/Catalog/Completeness/Checker/ChainedProductValueCompleteChecker.php (hunk 7 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Completeness/Checker/ChainedProductValueCompleteChecker.php]: seen IRL why it has been done like this

File: src/Pim/Component/Catalog/Completeness/Checker/ChainedProductValueCompleteChecker.php (hunk 12 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Completeness/Checker/ChainedProductValueCompleteChecker.php]: seen IRL why it has been done like this

File: src/Pim/Component/Catalog/Completeness/Checker/SimpleCompleteChecker.php (hunk 11 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Completeness/Checker/SimpleCompleteChecker.php]: done

File: src/Pim/Component/Catalog/Completeness/Checker/PriceCompleteChecker.php (hunk 10 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Completeness/Checker/PriceCompleteChecker.php]: done

**Pull request merged**